SpeechLM PR1 Modeling

jctian98 opened the pull request:
## What?
This PR adds the modeling files of SpeechLM modules, as well as the tokenizer definitions and `task/speechlm.py`
<!-- Please write what you changed. -->

## Why?
To merge code step-by-step. 
<!-- Please write why you changed. -->

## See also
#6050 
<!-- Write additional information if necessary (e.g., referecne, related PRs or Issues). -->

sw005320 commented on espnet2/bin/speechlm_inference_chat.py:
It is better to keep using `espnet2/bin/speechlm_inference.py` to follow the consistent binary naming.
See https://github.com/espnet/espnet/tree/master/espnet2/bin

The only exception is to have a variant in the inference algorithm (like `_k2` and `_maskctc`), and this case is different.

sw005320 replied:
So, `espnet2/bin/speechlm_inference_chat.py` --> `espnet2/bin/speechlm_inference.py`

jctian98 replied:
fixed

sw005320 commented on espnet2/speechlm/core_lm/ar_delay.py:
It's better to ask someone to check the details.
@wanchichen

sw005320 commented on espnet2/speechlm/core_lm/ar_parallel.py:
```diff
@@ -0,0 +1,109 @@
+        return x, targets, loss_mask
+
+    @torch.no_grad()
+    def inference(
```
Will you add this in the future?
If so, it's better to add TODO

jctian98 replied:
Add comment:

TODO(Jinchuan):
        We may not implement this inference code in the near future, as the parallel
        implementation would give poor performance. Anyone who is interested in 
        parallel should consider using ar_delay.py instead.

        Potentially, the parallel would be better aligned with the FSQ codec. Users
        who are interested in this direction could consider implementing the inference
        by themselves.

sw005320 commented on espnet2/speechlm/core_lm/ar_parallel.py:
```diff
@@ -0,0 +1,109 @@
+from espnet2.speechlm.inference_utils import AbsInferenceConfig
+
+
+class ARParallelLM(AbsCoreLM):
```
Is it used as an independent class, or is it only used for the ArDelayLM class?
If it is only used for the ArDelayLM class, we may merge this class with that.
If not, please explain it.

sw005320 commented on espnet2/speechlm/core_lm/ar_parallel.py:
It's better to ask someone to check the details.
@wanchichen

sw005320 commented on espnet2/speechlm/inference_utils.py:
```diff
@@ -0,0 +1,424 @@
+import torchaudio
+from kaldiio import WriteHelper
+
+from espnet2.speechlm.definitions import SPEECHLM_TASKS
```
Why is `SPEECHLM_TASKS` all upper-cases?
Any intention?
This breaks the naming convention.

jctian98 replied:
I want to emphasize that the SPEECHLM_TASKS are pre-defined hard codes shared in all SpeechLM modules, like MACROs. 
Would it be ok to keep it?

sw005320 commented on espnet2/speechlm/module/abs_transformer.py:
```diff
@@ -0,0 +1,38 @@
+# Copyright 2024 Jinchuan Tian
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+from abc import ABC, abstractmethod
```
what is `ABC`?
(sorry. simply I don't know this)

jctian98 replied:
abstract base class

sw005320 commented on espnet2/speechlm/module/transformer.py:
why did you remove this?

sw005320 commented on espnet2/speechlm/module/valle.py:
ditto
If there is a valid reason, please record it in the top of this PR

sw005320 reviewed (commented):
Please add detailed changes at the top of this PR.

Fhrozen commented:
This pull request introduces significant updates to the SpeechLM framework in the `espnet2` library, including refactoring the abstract base class, adding new architectures, and removing an outdated implementation. The changes aim to enhance modularity, simplify inference configurations, and support advanced language modeling techniques.

### Refactoring and Core Updates:
* [`espnet2/speechlm/core_lm/abs_core_lm.py`](diffhunk://#diff-40fbef862c275d4e73618819113de19b26f291045282cf54ef4c6e75886bb7caL7-R57): Refactored the abstract base class (`AbsCoreLM`) by removing the `SpeechLMInferenceOptions` dataclass and replacing it with `AbsInferenceConfig`. Updated method signatures to streamline arguments and added support for continuous feature tuples (`conti_feats`).

### New Architectures:
* [`espnet2/speechlm/core_lm/ar_parallel.py`](diffhunk://#diff-0123c16c3cecf942d6dfeb465112dcfb5a331c27ee24f4fc4ab23217bdd5350dR1-R109): Added `ARParallelLM`, a new auto-regressive language model supporting parallel interleave codec patterns. This includes modular continuous feature encoders and configurable embeddings.
* [`espnet2/speechlm/core_lm/ar_delay.py`](diffhunk://#diff-af6a51ed7d51e912ab695a93615b4156254f891e8bb10bef697421f001b707cdR1-R195): Introduced `ARDelayLM`, which implements a delay interleave pattern for training and inference, improving flexibility in sequence generation.

### Removal of Outdated Implementation:
* [`espnet2/speechlm/core_lm/ar_multiscale.py`](diffhunk://#diff-7f6d4726ac0b2b4a4ced7f2be71755626efe006f9d433cec878c757eb41af442L1-L267): Removed the `MultiScaleLM` implementation, which was based on the UniAudio architecture, to align with the updated framework design and focus on modularity.

jctian98 commented:
I should be able to have some updates in the following two days.

jctian98 commented on espnet2/speechlm/definitions.py:
```diff
@@ -95,7 +276,6 @@ def find_modality_type(self):  # Used in shell data preparation script
 #    previously can become incompatible. New tokens can be
 #    added - there are enough slots
-
 special_tokens = [
```
Add some explanation in the file, together with a link to the README.md

jctian98 commented on espnet2/speechlm/inference_utils.py:
```diff
@@ -0,0 +1,424 @@
+    return mask
+
+
+class TaskOrientedWriter:
```
comments added

jctian98 commented:
@sw005320  I've addressed the comments as above.

Since it's now targeting the espnet3 branch, the CI will not start automatically. I'll double-check the follow-up CI issues if someone could start the CI.

sw005320 commented:
@wanchichen, can you review this PR?

sw005320 commented:
@wanchichen, can you review this PR?

wanchichen commented on espnet2/speechlm/core_lm/ar_delay.py:
```diff
@@ -0,0 +1,227 @@
+
+        return retval, ret_loss_mask
+
+    def inverse_delay_interleave(
```
since delay-interleave and its reverse are deterministic, I think we should have a quick unit test to verify the input/output are properly shifted.

wanchichen commented on espnet2/speechlm/loss.py:
```diff
@@ -0,0 +1,168 @@
+                )
+                i += 1
+
+    def forward(self, hidden, targets, loss_mask):
```
also a unit test for this, since its susceptible to breaking as people add more modalities. I think we just need to check if the intervals and masking are correct for each loss

wanchichen commented:
overall the PR LGTM, the code is pretty well structured, just needs some tests to make sure it doesn't break in the future

wanchichen commented on espnet2/speechlm/core_lm/ar_parallel.py:
```diff
@@ -0,0 +1,119 @@
+        if share_emb:
+            self.lm_head.weight = self.emb.weight
+
+        self.head_emb = torch.nn.Embedding(12, transformer.d_model, padding_idx=0)
```
this is an embedding for each stream right? I think we should add it as a flag to `init()` and have a note that the input value must be larger than `nq`

jctian98 replied:
Fix it as `nq` rather than the hard code 12.

Since we always parse this `nq` from the model construction function, this nq should always be a valid value; otherwise, it would also trigger an error in line 84

`x = x.unsqueeze(2) + self.head_emb.weight.tile(1, 1, 1, 1)[:, :, : self.nq]`

sw005320 commented:
- Add TODO for your future actions (and do not forget them)
- Fix CI errors
- Fix issues pointed out by the Copilot

jctian98 commented on espnet2/speechlm/loss.py:
```diff
@@ -0,0 +1,168 @@
+                i += 1
+
+    def forward(self, hidden, targets, loss_mask):
+        # NOTE(Jinchuan): keep the weight on the correct device in the first forward.
```
We may not accept this comment since:

`# NOTE(Jinchuan): keep the weight on the correct device in the first forward.
        # We don't want to keep the weights registered as model parameters as they
        # should always be specified by external configurations.`

jctian98 commented on espnet2/speechlm/net_utils.py:
```diff
@@ -103,36 +76,104 @@ def install_hooks(layer: torch.nn.Module):
+
+    assert dec_emb.size(0) == len(conti_feats)
+
+    for b, conti_feat in enumerate(conti_feats):
```
Remove this function, not in use
